frontend/ui/widget/textboxwidget.lua

```diff
@@ -2186,4 +2176,247 @@ function TextBoxWidget:getSourceIndex(char_idx)
     end
 end
 
+function TextBoxWidget:getPosFromAbsPos(abs_pos)
+    local pos = Geom:new{
+        x = abs_pos.x - self.dimen.x,
+        y = abs_pos.y - self.dimen.y,
+    }
+
+    -- check if the coordinates are actually inside our area
+    if pos.x < 0 or pos.x >= self.dimen.w or pos.y < 0 or pos.y >= self.dimen.h then
+        return nil
+    end
+
+    return pos
+end
+
+function TextBoxWidget:getXtextHighlightIndices(start_x, start_y, end_x, end_y)
+    -- With xtext and fribidi, words may not be laid out in logical order,
+    -- so the left of a visual word may be its end in logical order,
+    -- and the right its start.
+    -- So, just find out charpos (text indice) of both points and
+    -- find word edges in the logical order text/charlist.
+    local sel_start_idx = self:getCharPosAtXY(start_x, start_y)
+    local sel_end_idx = self:getCharPosAtXY(end_x, end_y)
+    if not sel_start_idx or not sel_end_idx then
+        -- one or both hold points were out of text
+        return nil, nil, nil, nil
+    end
+    if sel_start_idx > sel_end_idx then -- re-order if needed
+        sel_start_idx, sel_end_idx = sel_end_idx, sel_start_idx
+    end
+    -- We get cursor positions, which can be after last char,
+    -- and that we need to correct. But if both positions are
+    -- after last char, the full selection is out of text.
+    if sel_start_idx > #self._xtext then -- Both are after last char
+        return nil, nil, nil, nil
+    end
+    if sel_end_idx > #self._xtext then -- Only end is after last char
+        sel_end_idx = #self._xtext
+    end
+
+    local start_line_num = math.ceil(start_y / self.line_height_px) + self.virtual_line_num - 1
+    local end_line_num = math.ceil(end_y / self.line_height_px) + self.virtual_line_num - 1
+    if start_line_num < 1 or end_line_num > #self.vertical_string_list then
+        return nil, nil, nil, nil
+    end
+
+    local word_start_idx, word_end_idx = self._xtext:getSelectedWordIndices(sel_start_idx, sel_end_idx, 50)
+    return word_start_idx, word_end_idx, start_line_num, end_line_num
+end
+
+function TextBoxWidget:getXtextHighlightRects(text_start_idx, text_end_idx, start_line_num, end_line_num)
+    local rects = {}
+
+    for line_num = start_line_num, end_line_num, 1 do
+        local line = self.vertical_string_list[line_num]
+        local draw_line = false
+        local line_x0 = 0
+        local line_x1 = 0
+
+        for _, xglyph in ipairs(line.xglyphs) do
+            if xglyph.text_index >= text_start_idx and (xglyph.text_index + xglyph.cluster_len) <= text_end_idx and (not xglyph.no_drawing) then
+                if draw_line then
+                    if xglyph.x0 < line_x0 then line_x0 = xglyph.x0 end
+                    if xglyph.x1 > line_x1 then line_x1 = xglyph.x1 end
+                else
+                    draw_line = true
+                    line_x0 = xglyph.x0
+                    line_x1 = xglyph.x1
+                end
+            end
+        end
```

Comment:
I think all this is new code - and I'm not keen on digging into that :) so trusting you.
I also don't know how this behave with BiDi text, and our text selection with fingers, and the resulting selected text in logical indices order - but I guess it's not impossible we should have multiple rectangles on a same line - your code would logically only emit one rectangle per line.
(Trying to see how this works in an Arabic with English bits EPUB, I don't get such multiple rectangles, so it's possible I screwed that recently...)
Anyway, just mentionning that; if the text selection was already crappy on BiDi text, let the highlighting be it too.

Reply:
Selecting within a single Arabic line does not work correctly. It is caused by koptinterface:getTextFromBoxes, I will fix that. But otherwise I have no idea how exactly it should behave. :)

Reply:
Just a sample of a single text selection on some mixed arabic/english search result:
![image](https://github.com/user-attachments/assets/eed9a216-7877-4539-ac76-829039013de6)